Provide a way to manage wp_navigation posts from wp-admin

anton-vlasenko opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
The goal of this PR is to disable block editor for wp_navigation type posts (because it doesn't work with wp_navigation posts) and provide a way to mange wp_navigation posts from wp-admin.
Fixes https://github.com/WordPress/gutenberg/issues/36036 (?).


## How has this been tested?
1. Create a new post using the block editor.
2. Add a new navigation block.
3. Create a new menu in your navigation block, add some menu items to it.
4. Save the post.
5. Go to `Appearance -> Navigation Menus` page.
6. Click on the post to edit it.
7.  Make sure that you cannot edit the post's content (i.e., your menu). Content editor should not be visible.

## Screenshots <!-- if applicable -->

![Screenshot 2021-11-03 at 15 37 31](https://user-images.githubusercontent.com/43744263/140081306-d9028138-fec0-4bf3-8d4e-cc5db63a31f9.png)

## Types of changes
New feature

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

anton-vlasenko committed:
Disable block editor for wp_navigation CPT (it doesn't work yet) so it can be managed using the standard UI.

draganescu commented:
I think we want to hide any ability to edit the content of the navigation CPT, as manually changing navigation block markup is prone to breaking it. Also, I like the simplicity of this approach if it works, no need for new API just a Gutenberg specific fix.

anton-vlasenko committed:
Disable ability to edit wp_navigation type posts because the post editor doesn't correctly work with wp_navigation posts.

anton-vlasenko committed:
Don't call get_posts when it's not needed.

anton-vlasenko committed:
Add unit tests for navigation.php.

anton-vlasenko commented:
@draganescu 
> I think we want to hide any ability to edit the content of the navigation CPT, as manually changing navigation block markup is prone to breaking it. Also, I like the simplicity of this approach if it works, no need for new API just a Gutenberg specific fix.

Thank you for the review.
The issues you mentioned should be fixed now. Now we cannot edit the content of a post (but we still can edit some other relevant data using the "Quick Edit" row action).
I'd appreciate it if you could review it.
cc @talldan

anton-vlasenko committed:
Fix code style.

talldan commented:
This looks great to me, thanks @anton-vlasenko!

The only remaining bit that would be great to tidy up is the bit of text here:
<img width="156" alt="Screenshot 2021-11-03 at 5 20 55 pm" src="https://user-images.githubusercontent.com/677833/140035621-7abf6415-5d4f-4ccf-86ec-87dc4833134b.png">

It'd be cool if it just said 'Navigation Menus'. I think this would just be a simple change to the strings defined in the custom post type definition:
https://github.com/WordPress/gutenberg/blob/40ff1e5014a823cb4b2aa49d5aa537fb108ed64d/lib/navigation.php#L402

The only part that confused me is that the Template Parts section has 'All Template Parts' defined for this `all_items` string, but as you can see the interface displays 'Template Parts':
https://github.com/WordPress/gutenberg/blob/40ff1e5014a823cb4b2aa49d5aa537fb108ed64d/lib/full-site-editing/template-parts.php#L25

So perhaps there's a different way to control this text that doesn't involve changing `all_items`.

anton-vlasenko committed:
Fix code style.

anton-vlasenko committed:
Fix code style.

anton-vlasenko committed:
Use get_post_type because it validates empty $post values.

talldan commented:
Oh, if there's a way to get rid of the 'Add new' button as well, that'd be great. It looks like that also opens the editor.

anton-vlasenko commented:
> Oh, if there's a way to get rid of the 'Add new' button as well, that'd be great. It looks like that also opens the editor.

Great catch. I've noticed it too.

draganescu commented:
Great work @anton-vlasenko 👏🏻 

@talldan this is what we get with this approach:

<img width="875" alt="Screen Shot 2021-11-03 at 15 14 18" src="https://user-images.githubusercontent.com/107534/140066536-d148284d-137c-4e35-844a-d9a26a5145da.png">

Aside from renaming, not much use for everything else there:

- publish date
- status
- password protecting or making private

Unless updating those changes the behavior of the block, I think this exploration will not lead us to far.

It feels to me like the whole idea of the #36036 is bound to be around using a block editor to properly edit a menu, while this approach using the "post edit screen" brings only one benefit and a lot of useless baggage.

anton-vlasenko committed:
1. Disable content editor for wp_navigation type posts.
Content editor cannot correctly edit them.
2. Add unit tests.

anton-vlasenko committed:
Fix menu entry.
Rename it to "Navigation Menus".

anton-vlasenko commented:
@draganescu @talldan 
It was not possible to remove the `Add New` button without modifying the core.
So I decided to remove the content editor.
I've updated the `How has this been tested?` section to reflect the latest changes.

anton-vlasenko commented:
@draganescu 
> It feels to me like the whole idea of the #36036 is bound to be around using a block editor to properly edit a menu, while this approach using the "post edit screen" brings only one benefit and a lot of useless baggage.

I see value in this PR (considering the changes I've just made).
It prevents users from editing content of `wp_navigation` posts using the `post edit` screen.
We can remove the `Quick Edit` row action if needed.

anton-vlasenko commented:
> So perhaps there's a different way to control this text that doesn't involve changing `all_items`.

Yes, there is a way.
We have to change `global $submenu` variable to achieve this.
At least this is how `Template Parts` CPT does it.
https://github.com/WordPress/gutenberg/blob/trunk/lib/full-site-editing/template-parts.php#L109

Fixed in https://github.com/WordPress/gutenberg/pull/36126/commits/39e9570f0936122ac21e9ba92a2d7dade6df5634

talldan commented:
> It feels to me like the whole idea of the #36036 is bound to be around using a block editor to properly edit a menu, while this approach using the "post edit screen" brings only one benefit and a lot of useless baggage.

I think bulk deleting is valuable right now, given we know the flow for creating menus isn't perfect.

And it'll be a good idea to iterate on providing a way to edit these menus via the Gutenberg plugin after 5.9.

The direction should ultimately be guided by https://github.com/WordPress/gutenberg/issues/29630 and future related issues to that. Right now it looks like that's moving towards using wp-admin for listing menus for 5.9.

spacedmonkey commented on lib/navigation.php:
```diff
@@ -418,13 +418,13 @@ function gutenberg_register_navigation_post_type() {
 		'map_meta_cap'          => true,
 		'rest_base'             => 'navigation',
-		'rest_controller_class' => 'WP_REST_Posts_Controller',
+		'rest_controller_class' => WP_REST_Posts_Controller::class,
```
Why change this?

anton-vlasenko replied:
IDEs like PHPStorm can understand what class is meant here if we use the `::class` keyword. This could be useful during refactoring or if we decide to use namespaces in the far future :). And it's just convenient to click on a class name to go to its definition.
I'm not aware of any modern PHP application that uses plain strings to specify class names. It just looks a little bit archaic to me.

draganescu replied:
I think this is good but, just thinking, `::class` returns the fully qualified class name resolution, it is not "the same" as the class name as a string, and if rest_controller_class does not expect a fully qualified class name we may get into trouble.

anton-vlasenko replied:
Yes, it does return a FQCN,  but since we don't use PHP namespaces in WordPress it equals to `'WP_REST_Posts_Controller'`. I've checked it.
And it's fail safe. Even If the `WP_REST_Posts_Controller` class is not loaded or defined, it will still return `'WP_REST_Posts_Controller'`.
I'm not insisting that we have to use `::class`, I just see no reasons not to use it.

draganescu replied:
What happens if we do introduce Namespaces? Well ... I assume that too many things will change for this to be a big problem?

anton-vlasenko replied:
If we do introduce namespaces, the FQCN will change to include the namespace.
E.g. instead of `'WP_REST_Posts_Controller'` it will return `'WordPress\REST\Controllers\Posts'` or something similar.
But if we introduce namespaces, we will need to know FQCN to load the correct class anyway because there could be several classes with the same name, i.e., multiple `Posts` classes in different namespaces.
This change shouldn't introduce any problems.

anton-vlasenko committed:
Enable editor support back.
We need to enable it back because we disable it to hide the content editor windows. This is required to avoid the incorrect state of wp_navigation CPT.

anton-vlasenko committed:
Fix the unit test.

draganescu commented on lib/navigation.php:
```diff
@@ -435,3 +435,80 @@ function gutenberg_register_navigation_post_type() {
+		return;
+	}
+
+	remove_post_type_support( $post_type, 'editor' );
```
So this has to hapen b/c if the post type definition itself does not support `editor` we get into the rest problem not allowing saving at all. That would be good to be added to the docblock of the function.

anton-vlasenko replied:
Great catch.
Fixed in 3bee0780ef210ad6704fa8010998a9a11b6b8d61

draganescu commented on lib/navigation.php:
```diff
@@ -435,3 +435,80 @@ function gutenberg_register_navigation_post_type() {
+add_action( 'edit_form_after_editor', 'gutenberg_enable_content_editor_for_navigation_post_type', 10, 1 );
+
+/**
+ * Fixes the label of the 'wp_navigation' admin menu entry.
```
It's unclear what exactly this fixes from the comment.

anton-vlasenko replied:
Thank you for the review!
I've renamed the method and updated the comment.
Fixed in baccf378c44a5b9ba977983f250f589716038075

draganescu commented on lib/navigation.php:
```diff
@@ -435,3 +435,80 @@ function gutenberg_register_navigation_post_type() {
+		return;
+	}
+	foreach ( $submenu['themes.php'] as $key => $submenu_entry ) {
+		if ( $post_type->labels->all_items === $submenu['themes.php'][ $key ][0] ) {
```
I think assigning `$submenu['themes.php'][ $key ][0]` to a variable that tells the reader what is at index `0` would be nice.

anton-vlasenko replied:
I've renamed the variables.
Please let me know if you think it's fine.
a6abfca97496fe9acbb5b7700a8fd3ee8a590853

draganescu commented on phpunit/navigation-test.php:
```diff
@@ -0,0 +1,94 @@
+<?php
+/**
+ * This class is supposed to test the functionality of the navigation.php
```
`is supposed to test` -> `tests`,
`of the navigation.php` -> `of block navigation `

anton-vlasenko replied:
Fixed in 38dbd8e19b65218ccebf46c07ef8515dc8ac667a

anton-vlasenko committed:
Refactor the gutenberg_fix_navigation_items_admin_menu_entry function.
The new name better describes the purpose of the function.

anton-vlasenko committed:
Rename variables to improve clarify of the code.

anton-vlasenko committed:
Fix description of the class.

anton-vlasenko committed:
Fix PHPDOC block. Now description better explains the purpose of this callback function.

talldan reviewed (approved):
Great stuff, LGTM.

Thanks for this, appreciate you finding all those niché core APIs for disabling features.

anton-vlasenko commented:
> Thanks for this, appreciate you finding all those niché core APIs for disabling features.

Thank you!
